.github/workflows/haskell-ci.yml

```diff
@@ -1,43 +1,80 @@
+      - master
 
 jobs:
+  cabal:
+    name: cabal ${{ matrix.os }} / ghc ${{ matrix.ghc }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os:
+          - ubuntu-latest
+        cabal: [3.4]
+        ghc:
+          - 8.6.5
+          - 8.8.4
+          - 8.10.7
+    steps:
+      - uses: actions/checkout@v2
+        if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'
+
+      - uses: haskell/actions/setup@v1
+        id: setup-haskell-cabal
+        name: Setup Haskell Cabal
+        with:
+          ghc-version: ${{ matrix.ghc }}
+          cabal-version: ${{ matrix.cabal }}
+
+      - uses: actions/cache@v2
+        name: Cache ~/.cabal/store
+        with:
+          path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
+          key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
+
+      - name: Build
+        run: cabal build --enable-tests --enable-benchmarks --write-ghc-environment-files=always
+
+      - name: Test
+        run: cabal test --enable-tests
+
   stack:
-    name: stack / resolver ${{ matrix.resolver }} 
+    name: ${{ matrix.stack-yaml }}
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        stack: ["2.7.3"]
-        resolver: ["lts-16.2"]
-        ghc: ["8.8.3"]
+        stack:
+          - 2.7.3
+        stack-yaml:
+          - stack-8.6.5.yaml
+          - stack-8.8.4.yaml
+          - stack-8.10.7.yaml
 
     steps:
-    - uses: actions/checkout@v2
-      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'
-
-    - uses: haskell/actions/setup@v1
-      name: Setup Haskell Stack
-      with:
-        ghc-version: ${{ matrix.ghc }}
-        stack-version: ${{ matrix.stack }}
-
-    - uses: actions/cache@v2.1.6
-      name: Cache ~/.stack
-      with:
-        path: ~/.stack
-        key: ${{ runner.os }}-${{ matrix.resolver }}-stack
-
-    - name: Install dependencies
-      run: |
-        stack build --test --bench --no-run-tests --no-run-benchmarks --only-dependencies
-    - name: Build
-      run: |
-        stack build --test --bench --no-run-tests --no-run-benchmarks
-    - name: Test
-      run: |
-        stack test
\ No newline at end of file
+      - uses: actions/checkout@v2
+        if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'
+
+      - uses: haskell/actions/setup@v1
+        name: Setup Haskell Stack
+        with:
+          stack-version: ${{ matrix.stack }}
+
+      - uses: actions/cache@v2.1.6
+        name: Cache ~/.stack
+        with:
+          path: ~/.stack
+          key: ${{ runner.os }}-${{ matrix.stack-yaml }}
+
+      - name: Install dependencies
+        run: |
+          stack build --stack-yaml ${{ matrix.stack-yaml }} --test --bench --no-run-tests --no-run-benchmarks --only-dependencies --system-ghc
```

Comment:
Since you're not setting up system ghc installation (you don't have `ghc-version: ${{ matrix.ghc }}` under haskell/action/setup in stack jobs)
it makes no sense to use system GHC.

Choose one of these 2 options:
1. install ghc in the setup by including ghc-version in the setup step + use --system-ghc flag 
OR
2. don't install ghc (so stack will install it itself when you run it for the first time) and remove --system-ghc flag.